packages/beacon-node/src/chain/seenCache/seenGossipBlockInput.ts

```diff
@@ -350,6 +361,100 @@ export class SeenGossipBlockInput {
     //   blockInputMeta: {pending: GossipedInputType.block, haveBlobs: blobsCache.size, expectedBlobs: null},
     // };
   }
+
+  private hasSampledDataColumns(dataColumnCache: DataColumnsCacheMap): boolean {
+    return (
+      dataColumnCache.size >= this.custodyConfig.sampledColumns.length &&
+      this.custodyConfig.sampledColumns.reduce((acc, columnIndex) => acc && dataColumnCache.has(columnIndex), true)
+    );
+  }
+
+  private async reconstructColumns(config: ChainForkConfig, blockRoot: RootHex) {
+    const blockCache = this.blockInputCache.get(blockRoot);
+    if (blockCache === undefined || blockCache.fork !== ForkName.fulu) {
+      return;
+    }
+
+    // If already have all columns, exit
+    if (
+      blockCache.cachedData &&
+      (blockCache.cachedData.fork !== ForkName.fulu ||
+        this.hasSampledDataColumns(blockCache.cachedData.dataColumnsCache))
+    ) {
+      return;
+    }
+
+    // Process KZG commitments into versioned hashes
+    let versionedHashes: Uint8Array[];
+
+    if (blockCache.block) {
+      const block = blockCache.block as fulu.SignedBeaconBlock;
+      versionedHashes = block.message.body.blobKzgCommitments.map(kzgCommitmentToVersionedHash);
+    } else if (blockCache.cachedData) {
+      const firstSidecar = blockCache.cachedData.dataColumnsCache.values().next().value;
+      if (!firstSidecar) {
+        return;
+      }
+
+      versionedHashes = firstSidecar.dataColumn.kzgCommitments.map(kzgCommitmentToVersionedHash);
+    } else {
+      throw new Error("blockInputCache missing both block and cachedData");
+    }
```

Comment:
Might be easier to set commitments and then check that some were found to catch the other error case in the comment above.  What do you think?

```suggestion
    let commitments: undefined | Uint8Array[];
    if (blockCache.block) {
      const block = blockCache.block as fulu.SignedBeaconBlock;
      commitments = block.message.body.blobKzgCommitments;
    } else if (blockCache.cachedData) {
      const firstSidecar = blockCache.cachedData.dataColumnsCache.values().next().value;
      commitments = firstSidecar?.dataColumn.kzgCommitments;
    }
    
    if (!commitments) {
      throw new Error("blockInputCache missing both block and cachedData");
    }
    
    // Process KZG commitments into versioned hashes
    const versionedHashes: Uint8Array[] = commitments.map(kzgCommitmentToVersionedHash);
```

Reply:
Pushed an update for this!